### PR TITLE
Extend RPC integration tests

### PR DESCRIPTION
**Pull request by tbenr**
Extends the abstract class to handle integration test on all milestones and transition combinations.
It systematically run RPC It might run more test combinations that is actually needed. The good thing is that we make sure nothing breaks when new milestone activates.

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRangeIntegrationTest.java [hunk 1]
```diff
@@ -155,38 +156,53 @@ void requestBlockBySlot_shouldReturnEmptyWhenBlockIsNotKnown() throws Exception
...
   }
 
-  @ParameterizedTest(name = "enableAltairLocally={0}, enableAltairRemotely={1}")
-  @MethodSource("altairVersioningOptions")
-  public void requestBlockBySlot_withDisparateVersionsEnabled_requestPhase0Blocks(
-      final boolean enableAltairLocally, final boolean enableAltairRemotely) throws Exception {
-    final Eth2Peer peer = createPeer(enableAltairLocally, enableAltairRemotely);
+  @ParameterizedTest(name = "{0} => {1}, nextSpecEnabledLocally={2}, nextSpecEnabledRemotely={3}")
+  @MethodSource("generateSpecTransitionWithCombinationParams")
+  public void requestBlockBySlot_withDisparateVersionsEnabled_requestBaseMilestoneBlocks(
```
Comment (zilm13) [inline networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRangeIntegrationTest.java]: it's byRange here

File: networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRangeIntegrationTest.java [hunk 2]
```diff
@@ -155,38 +156,53 @@ void requestBlockBySlot_shouldReturnEmptyWhenBlockIsNotKnown() throws Exception
...
     assertThat(response).containsExactly(block1.getBlock(), block2.getBlock());
   }
 
-  @ParameterizedTest(name = "enableAltairLocally={0}, enableAltairRemotely={1}")
-  @MethodSource("altairVersioningOptions")
-  public void requestBlockBySlot_withDisparateVersionsEnabled_requestAltairBlocks(
-      final boolean enableAltairLocally, final boolean enableAltairRemotely) throws Exception {
+  @ParameterizedTest(name = "{0} => {1}, nextSpecEnabledLocally={2}, nextSpecEnabledRemotely={3}")
+  @MethodSource("generateSpecTransitionWithCombinationParams")
+  public void requestBlockBySlot_withDisparateVersionsEnabled_requestNextSpecBlocks(
```
Comment (zilm13) [inline networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRangeIntegrationTest.java]: byRange this one too

File: networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRangeIntegrationTest.java [hunk 3]
```diff
@@ -217,39 +233,6 @@ public void requestBlockBySlot_withDisparateVersionsEnabled_requestAltairBlocks(
...
-
-  public static Stream<Arguments> altairVersioningOptions() {
-    return Stream.of(
-        Arguments.of(true, true),
-        Arguments.of(false, true),
-        Arguments.of(true, false),
-        Arguments.of(false, false));
-  }
-
   private List<SignedBeaconBlock> requestBlocks(final Eth2Peer peer)
```
Comment (zilm13) [inline networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRangeIntegrationTest.java]: I'd include byRange in this method name

File: networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRangeIntegrationTest.java [hunk 4]
```diff
@@ -155,38 +156,53 @@ void requestBlockBySlot_shouldReturnEmptyWhenBlockIsNotKnown() throws Exception
...
     final SignedBlockAndState block1 = peerStorage.chainUpdater().advanceChain();
     final SignedBlockAndState block2 = peerStorage.chainUpdater().advanceChain();
+
+    final Class<?> expectedBody = milestoneToBeaconBlockBodyClass(baseMilestone);
+
     peerStorage.chainUpdater().updateBestBlock(block2);
-    assertThat(block1.getBlock().getMessage().getBody()).isInstanceOf(BeaconBlockBodyPhase0.class);
-    assertThat(block1.getBlock().getMessage().getBody()).isInstanceOf(BeaconBlockBodyPhase0.class);
+    assertThat(block1.getBlock().getMessage().getBody()).isInstanceOf(expectedBody);
+    assertThat(block1.getBlock().getMessage().getBody()).isInstanceOf(expectedBody);
```
Comment (zilm13) [inline networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRangeIntegrationTest.java]: should be block2

File: networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRangeIntegrationTest.java (hunk 1 above)
Comment (zilm13) [inline networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRangeIntegrationTest.java]: as both blocks are from baseMilestone, deneb test is lost now, isn't it?

File: networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRootIntegrationTest.java [hunk 5]
```diff
@@ -179,39 +183,55 @@ void requestBlockByRoot_shouldReturnEmptyWhenBlockIsNotKnown() throws Exception
...
     final SignedBlockAndState block1 = peerStorage.chainUpdater().advanceChain();
     final SignedBlockAndState block2 = peerStorage.chainUpdater().advanceChain();
     peerStorage.chainUpdater().updateBestBlock(block2);
-    assertThat(block1.getBlock().getMessage().getBody()).isInstanceOf(BeaconBlockBodyPhase0.class);
-    assertThat(block1.getBlock().getMessage().getBody()).isInstanceOf(BeaconBlockBodyPhase0.class);
+
+    final Class<?> expectedBody = milestoneToBeaconBlockBodyClass(baseMilestone);
+
+    assertThat(block1.getBlock().getMessage().getBody()).isInstanceOf(expectedBody);
+    assertThat(block1.getBlock().getMessage().getBody()).isInstanceOf(expectedBody);
```
Comment (zilm13) [inline networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRootIntegrationTest.java]: should be block2

File: networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRootIntegrationTest.java [hunk 6]
```diff
@@ -179,39 +183,55 @@ void requestBlockByRoot_shouldReturnEmptyWhenBlockIsNotKnown() throws Exception
...
     assertThat(response).containsExactly(block1.getBlock(), block2.getBlock());
   }
 
-  @ParameterizedTest(name = "enableAltairLocally={0}, enableAltairRemotely={1}")
-  @MethodSource("altairVersioningOptions")
-  public void requestBlockBySlot_withDisparateVersionsEnabled_requestAltairBlocks(
-      final boolean enableAltairLocally, final boolean enableAltairRemotely) throws Exception {
+  @ParameterizedTest(name = "{0} => {1}, nextSpecEnabledLocally={2}, nextSpecEnabledRemotely={3}")
+  @MethodSource("generateSpecTransitionWithCombinationParams")
+  public void requestBlockBySlot_withDisparateVersionsEnabled_requestNextSpecBlocks(
```
Comment (zilm13) [inline networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRootIntegrationTest.java]: byRoot

File: networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRootIntegrationTest.java [hunk 7]
```diff
@@ -179,39 +183,55 @@ void requestBlockByRoot_shouldReturnEmptyWhenBlockIsNotKnown() throws Exception
...
-    assertThat(block1.getBlock().getMessage().getBody()).isInstanceOf(BeaconBlockBodyPhase0.class);
-    assertThat(block1.getBlock().getMessage().getBody()).isInstanceOf(BeaconBlockBodyPhase0.class);
+
+    final Class<?> expectedBody = milestoneToBeaconBlockBodyClass(baseMilestone);
+
+    assertThat(block1.getBlock().getMessage().getBody()).isInstanceOf(expectedBody);
+    assertThat(block1.getBlock().getMessage().getBody()).isInstanceOf(expectedBody);
 
     final List<SignedBeaconBlock> response =
         requestBlocks(peer, List.of(block1.getRoot(), block2.getRoot()));
```
Comment (zilm13) [inline networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRootIntegrationTest.java]: I'd rename called method to requestBlocksByRoot

File: networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRootIntegrationTest.java [hunk 8]
```diff
@@ -241,14 +261,6 @@ public void requestBlockBySlot_withDisparateVersionsEnabled_requestAltairBlocks(
     }
   }
```
Comment (zilm13) [inline networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRootIntegrationTest.java:264]: Why don't we have deneb test here?

**Review by zilm13 — CHANGES_REQUESTED** [review]

File: networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRangeIntegrationTest.java (hunk 1 above)
Comment (tbenr) [reply networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRangeIntegrationTest.java]: yeah.. haven't updated those names...

Comment (tbenr) [reply networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRangeIntegrationTest.java]: yes, deneb it is actually tested only on the other test `requestNextSpecBlocks` when both peers are in `nextSpec` enabled

Comment (tbenr) [reply networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRangeIntegrationTest.java]: it should be good enough to me

Comment (zilm13) [reply networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRangeIntegrationTest.java]: Ahh I see. Missed this. Great!

File: networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRootIntegrationTest.java (hunk 8 above)
Comment (zilm13) [reply networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRootIntegrationTest.java:264]: same here as in  BeaconBlocksByRangeIntegrationTest.java, checked in requestBlockByRoot_withDisparateVersionsEnabled_requestNextSpecBlocks, good

**Review by zilm13 — APPROVED** [review]
LGTM

**Pull request merged**